### PR TITLE
Wms t wms params

### PR DESCRIPTION
**Pull request by lucianpls**
## What does this PR do?

WMS: tWMS, improves usability with NASA GIBS services

## What are related issues/pull requests?

Add a couple of open options to the tiled WMS minidriver, to make it more user friendly to connect to NASA GIBS services.
-oo TiledGroupName=<Name>
  When a tiledWMS handle file does not contain a TiledGroupName node, this open option can be used to pick one.
  Example: -oo TiledGroupname="MODIS Terra CorrectedReflectance TrueColor tileset"
-oo Change=<key>:<value>
  When a tiledWMS service offers declares parameters using the Change node, this open option can be used to provide a value.
  GIBS usually allows a **time** parameter, with allowed values being ISO9601 formatted dates
  Example: -oo Change=time:2021-01-30

The open options work with the WMS metadriver also, where the TiledGroupName becomes a substring filter.
This command will show as subdatasets all the datasets that contain the word **thermal**
  gdalinfo -oo TiledGroupName="thermal" https://gibs.earthdata.nasa.gov/twms/epsg4326/best/twms.cgi?

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Add documentation
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

File: gdal/frmts/wms/minidriver_tiled_wms.cpp [hunk 1]
```diff
@@ -237,26 +265,22 @@ static void FindChangePattern(char *cdata, char **substs, char **keys, CPLString
                 // But is the match for the key position?
                 char *found_key = nullptr;
                 const char *found_value = CPLParseNameValue(substs[sub_number], &found_key);
-                if (found_key != nullptr && EQUAL(found_key, key))
-                {  // Should exits in the request
+                if (found_key != nullptr && EQUAL(found_key, key)) {
+                    // Should exist in the request
```
Comment (mloskot) [inline gdal/frmts/wms/minidriver_tiled_wms.cpp:269]: It is
- mixture of code style change with functional change (in other lines)
- code style change to a style different than the majority of GDAL codebase

and both are discouraged and undesired.

**Review by mloskot — CHANGES_REQUESTED** [review]

File: gdal/frmts/wms/minidriver_tiled_wms.cpp (hunk 1 above)
Comment (lucianpls) [reply gdal/frmts/wms/minidriver_tiled_wms.cpp]: OK

File: gdal/frmts/wms/wmsutils.cpp [hunk 2]
```diff
@@ -28,24 +28,26 @@
 
 #include "wmsdriver.h"
 
+CPL_CVSID("$Id$")
+
```
Comment (mloskot) [inline gdal/frmts/wms/wmsutils.cpp]: https://github.com/OSGeo/gdal/pull/2503#issuecomment-626307246

Comment (lucianpls) [reply gdal/frmts/wms/wmsutils.cpp]: Yes, but I didn't want to show this file as modified, it was just busy fingers.

File: gdal/frmts/wms/wmsmetadataset.cpp [hunk 3]
```diff
@@ -662,33 +676,52 @@ void GDALWMSMetaDataset::AddTiledSubDataset(const char* pszTiledGroupName,
 /************************************************************************/
 /*                     AnalyzeGetTileServiceRecurse()                   */
 /************************************************************************/
-
-void GDALWMSMetaDataset::AnalyzeGetTileServiceRecurse(CPLXMLNode* psXML)
+// tiledWMS only
+void GDALWMSMetaDataset::AnalyzeGetTileServiceRecurse(CPLXMLNode* psXML, GDALOpenInfo * poOpenInfo)
 {
+    // Only list tiled groups that contain the string in the open option TiledGroupName, if given
+    char **OpenOptions = poOpenInfo ? poOpenInfo->papszOpenOptions : nullptr;
```
Comment (mloskot) [inline gdal/frmts/wms/wmsmetadataset.cpp:683]: Variables should read in lowercase, e.g. `papszOpenOptions`

File: gdal/frmts/wms/wmsmetadataset.cpp [hunk 4]
```diff
@@ -662,33 +676,52 @@ void GDALWMSMetaDataset::AddTiledSubDataset(const char* pszTiledGroupName,
...
+    CPLString match(CSLFetchNameValueDef(OpenOptions, "TiledGroupName",""));
+    match.toupper();
+    // Also pass the change patterns, if provided
+    char **changes = CSLFetchNameValueMultiple(OpenOptions, "Change");
```
Comment (mloskot) [inline gdal/frmts/wms/wmsmetadataset.cpp:687]: `papszChanges`

File: gdal/frmts/wms/wmsmetadataset.cpp [hunk 5]
```diff
@@ -662,33 +676,52 @@ void GDALWMSMetaDataset::AddTiledSubDataset(const char* pszTiledGroupName,
...
-            if (pszName)
-                AddTiledSubDataset(pszName, pszTitle);
+            if (!name.empty())
+            {
+                if (match.empty())
+                {
+                    AddTiledSubDataset(name, pszTitle, changes);
+                }
+                else {
+                    CPLString NAME(name);
```
Comment (mloskot) [inline gdal/frmts/wms/wmsmetadataset.cpp:704]: `osName`

File: gdal/frmts/wms/wmsmetadataset.cpp [hunk 6]
```diff
@@ -662,33 +676,52 @@ void GDALWMSMetaDataset::AddTiledSubDataset(const char* pszTiledGroupName,
...
+    char **changes = CSLFetchNameValueMultiple(OpenOptions, "Change");
+
     CPLXMLNode* psIter = psXML->psChild;
     for(; psIter != nullptr; psIter = psIter->psNext)
     {
         if (psIter->eType == CXT_Element &&
             EQUAL(psIter->pszValue, "TiledGroup"))
         {
-            const char* pszName = CPLGetXMLValue(psIter, "Name", nullptr);
+            CPLString name(CPLGetXMLValue(psIter, "Name", ""));
```
Comment (mloskot) [inline gdal/frmts/wms/wmsmetadataset.cpp:695]: `osName`

File: gdal/frmts/wms/wmsmetadataset.cpp [hunk 7]
```diff
@@ -662,33 +676,52 @@ void GDALWMSMetaDataset::AddTiledSubDataset(const char* pszTiledGroupName,
...
         }
-        else if (psIter->eType == CXT_Element &&
-            EQUAL(psIter->pszValue, "TiledGroups"))
+        else if (psIter->eType == CXT_Element && EQUAL(psIter->pszValue, "TiledGroups"))
         {
-            AnalyzeGetTileServiceRecurse(psIter);
+            AnalyzeGetTileServiceRecurse(psIter, poOpenInfo);
         }
     }
+    if (changes)
```
Comment (mloskot) [inline gdal/frmts/wms/wmsmetadataset.cpp:716]: Redundant.
`CPLFree` and `VSIFree` no-op for `nullptr`

File: gdal/frmts/wms/wmsmetadataset.cpp [hunk 8]
```diff
@@ -625,14 +623,30 @@ GDALDataset* GDALWMSMetaDataset::AnalyzeGetCapabilities(CPLXMLNode* psXML,
...
+
+    if (changes)
+    {
+        for (int i = 0; changes[i] != nullptr; i++)
+        {
+            char* key = nullptr;
+            const char* value = CPLParseNameValue(changes[i], &key);
+            if (value != nullptr && key != nullptr)
+                osSubdatasetName += CPLSPrintf("<Change key=\"${%s}\">%s</Change>", key, value);
+            if (key) CPLFree(key);
```
Comment (mloskot) [inline gdal/frmts/wms/wmsmetadataset.cpp:645]: `if (key)` is redundant
`CPLFree` and `VSIFree` no-op for `nullptr`

File: gdal/frmts/wms/wmsmetadataset.cpp [hunk 9]
```diff
@@ -625,14 +623,30 @@ GDALDataset* GDALWMSMetaDataset::AnalyzeGetCapabilities(CPLXMLNode* psXML,
 /*                          AddTiledSubDataset()                        */
 /************************************************************************/
 
+// tiledWMS only
 void GDALWMSMetaDataset::AddTiledSubDataset(const char* pszTiledGroupName,
-                                            const char* pszTitle)
+                                            const char* pszTitle,
+                                            const char* const* changes)
```
Comment (mloskot) [inline gdal/frmts/wms/wmsmetadataset.cpp:629]: `s/changes/papszChanges/`

**Review by mloskot — CHANGES_REQUESTED** [review]

File: gdal/frmts/wms/wmsutils.cpp (hunk 2 above)
Comment (mloskot) [reply gdal/frmts/wms/wmsutils.cpp]: Gotcha

**Comment by mloskot** [comment]
It would be nice if the enhancement is covered with a test case, if possible.

**Comment by lucianpls** [comment]
> 
> 
> It would be nice if the enhancement is covered with a test case, if possible.

I'm waiting for GIBS to fix an SSL issue they are having currently.  The workaround is too cumbersome.

File: gdal/frmts/wms/wmsmetadataset.cpp (hunk 7 above)
Comment (lucianpls) [reply gdal/frmts/wms/wmsmetadataset.cpp:716]: Doesn't hurt either and it makes the intention clear.

Comment (mloskot) [reply gdal/frmts/wms/wmsmetadataset.cpp:716]: No such intention should ever exist here. It's bad habbit, it's cruft, [incorrect code potential](https://isocpp.org/wiki/faq/freestore-mgmt#delete-handles-null).

Comment (lucianpls) [reply gdal/frmts/wms/wmsmetadataset.cpp]: Fine, even though I disagree.
I'm not convinced that it's a no-op in all cases, even though in this particular case it doesn't make a difference.

Comment (lucianpls) [reply gdal/frmts/wms/wmsmetadataset.cpp]: Check this out, the default implementation of VSIFree : cpl_vsisimple.cpp line 872
```
if (pData != nullptr)
  free( pData );
```

File: gdal/frmts/wms/minidriver_tiled_wms.cpp [hunk 10]
```diff
@@ -29,19 +29,47 @@
  * DEALINGS IN THE SOFTWARE.
  ****************************************************************************/
 
+
+ //
+ // Also known as the OnEarth tile protocol
+ //
+ // A few open options are supported by tiled WMS
```
Comment (rouault) [inline gdal/frmts/wms/minidriver_tiled_wms.cpp:36]: should rather be moved to wms.rst

**Pull request merged**